### ISSUE TITLE
Pre-R6 — prove a physically external Coordinator over real compute Nodes

### ISSUE THREAD

**Issue by ezutfen** [Zutfen-LLC/inferswarm#67]
## Objective

Prove that the InferSwarm **Coordinator is genuinely a replaceable control-plane role rather than an implicit compute-host role** by moving request ingress, planning, frozen Execution Plan authority, epoch/session coordination, and result commitment onto the dedicated CPU-only `inferswarm00` VM while all correctness-bearing model materialization and inference execution remain on `inferswarm01` and/or `inferswarm03`.

This is a bounded architecture/integration prerequisite for R6. It is **not** Coordinator HA, election, consensus, production RPC design, or Gemma implementation.

The question is:

> Can an ordinary known-good serving request be authoritatively coordinated from a host with no Torch, CUDA, NVIDIA runtime, native FreeToken execution extensions, or model weights while real execution occurs entirely on remote Compute Nodes?

## Accepted prerequisites

Treat as completed and immutable:

- R5B: `R5B_PLAN_EPOCH_RECOVERY_PASS`
- Pre-R6 integration: `PRE_R6_INTEGRATION_REQUALIFICATION_PASS`
- accepted protected FreeToken research head:
  `8cfcda4c44065ceba4230dc548a12696093d5177`
- accepted InferSwarm coordination/docs head:
  `ab831da22b1c75ee1e328ca066598f0da3eed5cb`

R6 / #65 remains blocked until this external-Coordinator proof is accepted.

Do not modify accepted R4, R5A, R5B, Pre-R5, or Pre-R6 evidence.

## Provisioned Coordinator baseline

`inferswarm00` has already been provisioned specifically for this role.

Frozen starting facts to re-verify before canonical evidence:

- hostname: `inferswarm00`
- LAN: `10.0.0.206/24` on `ens18`, MTU 1500
- Debian GNU/Linux 13, Python 3.13.5
- KVM VM, 4 vCPU, ~8 GiB RAM
- no NVIDIA driver
- no CUDA
- no `nvidia-smi`
- no `/dev/nvidia*`
- no Torch
- no Triton
- no FreeToken native compute extensions
- no model weights
- FreeToken checkout at accepted research head `8cfcda4c...`
- InferSwarm checkout at `ab831da2...`
- generic R3 planner, R5A frozen-plan/reconciliation, and R5B epoch-controller modules already import and execute CPU-only
- focused generic control-plane tests passed 65/65 during provisioning

The local provisioning record is `/srv/inferswarm/COORDINATOR_PROVISIONING.md`; it is not repository evidence unless deliberately copied into a new retained record for this issue.

Do not install GPU/runtime/model dependencies on `inferswarm00` to make this issue pass.

## Physical topology

Freeze actual identities and current state before canonical execution.

### Coordinator

`inferswarm00`

Coordinator/control-plane only.

### Compute Nodes

Use the current accepted RTX 3060 proving Nodes:

- `inferswarm01`
- `inferswarm03`

Freeze current:

- Node identities;
- GPU UUID/BDF/VRAM identities;
- runtime/software versions;
- checkpoint identity/hashes;
- memory/headroom;
- current network paths and negotiated state;
- current integrity/trust status.

Do not blindly reuse R5B measurements as though the environment cannot have changed.

Prefer the ordinary LAN path where available. If the canonical run uses Tailscale or another route, record that explicitly and do not label it as the direct-LAN R4/R5 path.

## Architectural invariant under test

The accepted Fabric Doctrine states that:

- a `Swarm` is durable and is not synonymous with one machine;
- a `Coordinator` is a replaceable control-plane role;
- a `Node` is one physical resource/runtime authority;
- workers/executors are runtime/Execution Plan constructs.

This issue physically tests that separation.

A passing implementation must not turn the Coordinator into a remote CUDA orchestrator or require it to host a redundant model materialization.

## Required separation

For the canonical arm, `inferswarm00` must own or authoritatively mediate:

- ordinary request ingress;
- session identity;
- trusted resource snapshot consumed by planning;
- model/workload strategy inputs needed for planning;
- generic planner execution;
- selected candidate and frozen Execution Plan identity;
- plan epoch/generation authority;
- realization authorization;
- committed-output boundary/accounting;
- result acceptance/fencing;
- client-visible output commitment.

`inferswarm01` / `inferswarm03` must own, as applicable:

- local GPU discovery and integrity validation;
- checkpoint/model materialization;
- CUDA/backend initialization;
- backend-native execution;
- GPU/device-local runtime objects;
- local mutable runtime state assigned by the strategy;
- local execution-process lifecycle;
- observed realization data returned for reconciliation.

No correctness-bearing model execution may occur on `inferswarm00`.

## Remote realization/execution seam

The accepted epoch controller currently receives an in-process `realizer` callable returning a `RealizedStaticPlan` / runtime object. External Coordinator operation requires a research-internal remote equivalent.

Implement the **minimum semantic remote realization/execution contract needed for this proof**.

Conceptually:

```text
ordinary client request
        |
        v
inferswarm00
  Coordinator
  planner / epoch authority
        |
        | frozen Execution Plan
        | realization + execution authorization
        v
Node-local execution agent(s)
  inferswarm01 / inferswarm03
        |
        | backend-specific realization
        v
FreeToken / CUDA / GPU
        |
        | attributed result + observation
        v
inferswarm00
  authority/fencing/commit
        |
        v
client-visible output
```

The Coordinator should speak in Execution Plan / session / epoch / operation semantics. Backend-specific CUDA/process details remain behind the Node-local execution boundary.

Do **not** freeze a public Node-agent API or production wire protocol.

## Research transport

Ordinary TCP is acceptable for this gate.

Use a simple bounded research-internal encoding with explicit framing and fail-closed parsing. JSON plus length framing, msgpack, or another inspectable encoding is acceptable if justified.

Do not introduce a generalized service mesh, gRPC platform, distributed database, or consensus system merely to complete this proof.

The canonical listening ports must be explicitly recorded and scoped to the test LAN/control path.

## Required authority and attribution

Every correctness-bearing remote request/result must be attributable to enough immutable identity to prevent cross-session, cross-plan, or retired-epoch acceptance.

At minimum include or cryptographically bind as applicable:

- protocol/schema version;
- Swarm/test-scope identity;
- session/request identity;
- Coordinator-authorized epoch/generation;
- frozen Execution Plan digest;
- operation identity;
- expected execution/commit position where applicable;
- participant/role identity;
- payload checksum/digest for strategy-semantic tensors/results where exposed.

The Coordinator must reject results that do not match the currently authorized session/epoch/plan/position.

The Node agent must reject malformed, unsupported, stale, or internally inconsistent realization requests rather than guessing.

## Minimum fencing proof

Retain at least one controlled negative arm proving that an invalid remote result cannot mutate current output/accounting.

Preferred arm:

1. complete or supersede one real authorized operation;
2. inject a controlled duplicate/stale result carrying an old operation identity or intentionally wrong plan/epoch identity;
3. route it through the same Coordinator acceptance path;
4. prove mechanical rejection;
5. prove committed token/output/accounting state remains unchanged.

Do not claim Coordinator-term fencing or Coordinator HA; there is still exactly one Coordinator in this issue.

## Model/workload

Use the accepted known-good Qwen3.6 proving model, not Gemma.

Checkpoint identity should remain the exact supported Qwen3.6 NVFP4 checkpoint used by the accepted current research line unless preflight proves a necessary dependency-scoped adjustment:

`nvidia/Qwen3.6-35B-A3B-NVFP4`

Freeze the exact revision/checkpoint identity before canonical evidence.

Use a small deterministic known-good workload derived from accepted R5A/R5B/Pre-R6 serving evidence.

Prefer the established exact-token reference beginning:

`[9764, 393, ...]`

The canonical proof need only generate enough tokens to prove request ingress, remote execution, Coordinator-side commit, and fencing. It does not need the full expensive R5B lifecycle.

## Planner requirement

Planning must physically execute on `inferswarm00` using the accepted generic planner seam.

The Coordinator must freeze the resource snapshot, planner decision, and selected Execution Plan before heavyweight realization.

The planner must not be replaced with a manually hard-coded `use inferswarm01/03` command merely because this is a topology proof.

A controlled evidence-collection override is acceptable only under the existing semantics and must remain visibly distinct from automatic selection.

## Realization/reconciliation requirement

The selected plan must be realized by the remote Node execution substrate and observed state returned to `inferswarm00`.

The Coordinator must run the accepted reconciliation logic against the frozen plan.

A passing arm requires:

- exact intended participant/resource identities;
- complete expected state/materialization coverage for the reused Qwen plan;
- no silent resource substitution;
- no unexpected fallback;
- no unplanned source access;
- no unexplained persistent host mirror;
- no unplanned steady-state model-state movement;
- clean staging/reclamation accounting appropriate to the selected plan.

## Coordinator purity invariant

For the entire canonical run, retain evidence that `inferswarm00` has:

- no NVIDIA driver/runtime dependency introduced for InferSwarm;
- no `nvidia-smi` requirement;
- no CUDA installation;
- no Torch requirement unless an explicit CPU-only tokenizer/control dependency is independently justified and documented;
- no Triton;
- no FreeToken native GPU execution extension;
- no model-weight materialization;
- no `/dev/nvidia*` devices.

If the implementation requires one of these merely to coordinate remote execution, treat that as a layering failure and refactor the seam rather than installing the dependency silently.

Small tokenizer/config metadata may be retained on the Coordinator only if the ordinary serving path genuinely requires them. Account for such artifacts separately from model weights.

## Ordinary serving proof

The canonical request must enter through `inferswarm00` and complete through the ordinary OpenAI-compatible serving path or the narrowest faithful adaptation of the existing `/v1/chat/completions` serving waist.

Required conceptual path:

```text
/v1/chat/completions on inferswarm00
  -> request/session setup
  -> strategy legal candidates
  -> generic planner on inferswarm00
  -> frozen Execution Plan
  -> remote Node realization
  -> remote backend-native execution
  -> attributed result returned to inferswarm00
  -> Coordinator authority/fencing check
  -> committed token/output
  -> HTTP response
```

Do not satisfy this issue with a benchmark-only script that bypasses ordinary request/commit semantics.

## Correctness

Freeze the exact workload and reference before interpreting canonical results.

Require at minimum:

- exact prompt token IDs;
- deterministic sampling settings;
- exact generated-token equality to the accepted reference for the retained scope;
- zero duplicate/missing/reordered committed tokens;
- zero NaN/Inf where logits/tensors are retained;
- semantic payload/boundary checksum validation where exposed;
- exact session identity;
- exact plan/epoch attribution;
- clean Coordinator-side committed-output accounting.

Do not weaken thresholds after observing results.

## Shutdown/reclamation

After the canonical request:

- close/reclaim remote epoch runtimes;
- stop research Node-agent processes started specifically for this proof;
- prove no unexpected GPU execution processes remain;
- prove Coordinator-side session/runtime state reaches the expected terminal state;
- retain any deliberately persistent agent daemon separately if one is intentionally part of the tested substrate and document why.

Prefer bounded per-proof processes over prematurely creating production daemons.

## Tests and regressions

At minimum add/run:

- remote protocol framing/validation tests;
- Node-agent authorization/validation tests;
- Coordinator remote-realizer tests;
- session/plan/epoch mismatch rejection tests;
- stale/duplicate result fencing test;
- reconnect/remote error fail-closed tests appropriate to the minimal protocol;
- existing generic planner tests;
- existing frozen-plan/reconciliation tests;
- existing epoch-controller tests;
- current R1-R5B research regressions;
- InferSwarm benchmark regressions;
- server regressions.

Run relevant tests on the CPU-only Coordinator environment where possible to prove control-plane purity.

Classify anomalies experimentally.

## Failure semantics

For this gate, if the Coordinator loses contact with a required Node during an active operation, it may fail the request/session closed unless the existing accepted epoch recovery machinery is deliberately exercised.

Full Node-loss recovery is already established by R5B and need not be re-proven here.

Coordinator loss/replacement is explicitly out of scope.

Do not fabricate continuation after an uncertain remote outcome. If the Coordinator cannot determine whether a correctness-bearing operation committed, fail closed under the narrowest affected scope.

## Evidence retention

Create a new retained directory such as:

`docs/inferswarm_external_coordinator/`

Retain at minimum:

- methodology frozen before canonical physical results;
- accepted base SHA;
- Coordinator provisioning/environment census;
- compute-Node preflight;
- network/path census;
- remote protocol schema/version and rationale;
- planner snapshot/decision;
- frozen Execution Plan;
- realization authorization records;
- Node observations/reconciliation;
- ordinary HTTP request/response evidence;
- exact generated-token comparator;
- authority/commit records;
- controlled stale/mismatch rejection arm;
- Coordinator purity evidence;
- RAM/VRAM/staging/source/mirror/movement accounting;
- shutdown/reclamation evidence;
- regression summary;
- anomalies;
- manifest/hashes;
- explicit non-claims.

Do not edit accepted historical evidence directories.

## Documentation synchronization

Update top-level InferSwarm status so this issue is the current prerequisite and #65/R6 remains blocked until acceptance.

After this issue passes and its FreeToken work is merged, synchronize the roadmap again to mark this gate complete and #65 current/unblocked with the exact new research head.

Do not rewrite historical R5 measurements.

## Explicit non-goals

This issue does **not** require or claim:

- Coordinator high availability;
- Coordinator election;
- Raft/Paxos/consensus;
- multiple active Coordinators;
- Coordinator-term fencing;
- automatic Coordinator takeover;
- zero-downtime Coordinator failover;
- production Node-agent daemon/API;
- public wire protocol;
- TLS/PKI design beyond what is necessary for the bounded trusted-LAN research proof;
- Gemma 4 support;
- R6 architecture falsification;
- new Qwen placement/performance claims;
- optimized remote-control latency.

## Passing condition

Declare:

`EXTERNAL_COORDINATOR_SEPARATION_PASS`

only when retained physical evidence proves all of the following:

- exact accepted Pre-R6 FreeToken base is used;
- `inferswarm00` remains CPU-only and free of local model/GPU execution dependencies;
- ordinary request ingress and authoritative session/plan/epoch control occur on `inferswarm00`;
- generic planning physically occurs on `inferswarm00`;
- selected Execution Plan is frozen before heavyweight realization;
- all model materialization and correctness-bearing inference execution occur on remote compute Node(s);
- remote realization returns enough observed state for clean Coordinator-side reconciliation;
- exact known-good Qwen output is preserved;
- Coordinator commits output only after validating current session/plan/epoch/operation authority;
- a controlled stale/mismatched result is mechanically rejected without changing committed state/accounting;
- residency/staging/source/mirror/movement invariants remain clean;
- shutdown/reclamation is complete;
- applicable regressions pass or deviations are experimentally classified;
- no Coordinator HA, public protocol, production daemon, Gemma, or R6 claim is made.

If a clean external Coordinator requires changing the generic internal seam, make the smallest model-independent refactor supported by the evidence and retain it explicitly.

If the existing architecture cannot support external coordination without introducing GPU/model coupling onto `inferswarm00`, report the exact blocker rather than manufacturing the pass token.

**Comment by ezutfen** [comment]
Accepted and completed.

Final disposition: `EXTERNAL_COORDINATOR_SEPARATION_PASS`.

Canonical provenance:
- FreeToken base entering gate: `8cfcda4c44065ceba4230dc548a12696093d5177`
- final physical producer: `2bcf33f6d6e5dc9fc5c13e37e7110833cbad0fcd`
- accepted retained evidence head: `603a63bf44728814d3182191dcd87e21229e5370`
- normal FreeToken PR #24 merge / new protected `inferswarm-research` head: `84ebd2b7ae56c60292f7b9c7ca256f41f64d8b11`

The earlier physical producers `df8a429e9110…` and `84e531971d6c…` remain retained as explicitly superseded evidence and are not the canonical proof.

InferSwarm roadmap/status sync merged via PR #68 as `e78be22de1814b0ed1b95a5a64366541ca0272a0`.

#65 / R6 is now current and unblocked on exact FreeToken base `84ebd2b7ae56c60292f7b9c7ca256f41f64d8b11`.
